MediaBrowser.Providers/Lyric/LyricScheduledTask.cs

```diff
@@ -0,0 +1,171 @@
+    private readonly ILogger<LyricDownloadTask> _logger;
+    private readonly ILocalizationManager _localizationManager;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="LyricDownloadTask"/> class.
+    /// </summary>
+    /// <param name="libraryManager">Instance of the <see cref="ILibraryManager"/> interface.</param>
+    /// <param name="lyricManager">Instance of the <see cref="ILyricManager"/> interface.</param>
+    /// <param name="logger">Instance of the <see cref="ILogger{DownloaderScheduledTask}"/> interface.</param>
+    /// <param name="localizationManager">Instance of the <see cref="ILocalizationManager"/> interface.</param>
+    public LyricDownloadTask(
+        ILibraryManager libraryManager,
+        ILyricManager lyricManager,
+        ILogger<LyricDownloadTask> logger,
+        ILocalizationManager localizationManager)
+    {
+        _libraryManager = libraryManager;
+        _lyricManager = lyricManager;
+        _logger = logger;
+        _localizationManager = localizationManager;
+    }
+
+    /// <inheritdoc />
+    public string Name => _localizationManager.GetLocalizedString("TaskDownloadMissingLyrics");
+
+    /// <inheritdoc />
+    public string Key => "DownloadLrcLibLyrics";
+
+    /// <inheritdoc />
+    public string Description => _localizationManager.GetLocalizedString("TaskDownloadMissingLyricsDescription");
+
+    /// <inheritdoc />
+    public string Category => _localizationManager.GetLocalizedString("TasksLibraryCategory");
+
+    /// <inheritdoc />
+    public async Task ExecuteAsync(IProgress<double> progress, CancellationToken cancellationToken)
+    {
+        var totalCount = _libraryManager.GetCount(new InternalItemsQuery
+        {
+            Recursive = true,
+            IsVirtualItem = false,
+            IncludeItemTypes = _itemKinds,
+            DtoOptions = _dtoOptions,
+            MediaTypes = _mediaTypes,
+            SourceTypes = _sourceTypes
+        });
+
+        var completed = 0;
+
+        foreach (var library in _libraryManager.RootFolder.Children.ToList())
+        {
+            var libraryOptions = _libraryManager.GetLibraryOptions(library);
+            var itemQuery = new InternalItemsQuery
+            {
+                Recursive = true,
+                IsVirtualItem = false,
+                IncludeItemTypes = _itemKinds,
+                DtoOptions = _dtoOptions,
+                MediaTypes = _mediaTypes,
+                SourceTypes = _sourceTypes,
+                Limit = QueryPageLimit,
+                Parent = library
+            };
+
+            int previousCount;
+            var startIndex = 0;
+            do
+            {
+                itemQuery.StartIndex = startIndex;
+                var audioItems = _libraryManager.GetItemList(itemQuery);
+
+                foreach (var audioItem in audioItems.OfType<Audio>())
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
```

Comment:
i would prefer to abort gracefully here instead of throwing

Reply:
The `ScheduledTaskWorker` sets the status based on how the scheduled task completes so right now this is how we specify that the task has been cancelled instead of completed successfully.